internal/mock/sqladmin.go

```diff
@@ -0,0 +1,231 @@
+//
+// https://cloud.google.com/sql/docs/mysql/admin-api/rest/v1beta4/instances/get
+func InstanceGetSuccess(i CloudSQLInst, ct int) *Request {
+	// Turn instance keys/certs into PEM encoded versions needed for response
+	certBytes, err := x509.CreateCertificate(
+		rand.Reader, i.cert, i.cert, &i.privKey.PublicKey, i.privKey)
+	if err != nil {
+		panic(err)
+	}
+	certPEM := new(bytes.Buffer)
+	pem.Encode(certPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certBytes,
+	})
+
+	db := &sqladmin.DatabaseInstance{
+		BackendType:     "SECOND_GEN",
+		ConnectionName:  fmt.Sprintf("%s:%s:%s", i.project, i.region, i.name),
+		DatabaseVersion: i.dbVersion,
+		Project:         i.project,
+		Region:          i.region,
+		Name:            i.name,
+		IpAddresses: []*sqladmin.IpMapping{
+			{
+				IpAddress: "127.0.0.1",
+				Type:      "PRIMARY",
+			},
+		},
+		ServerCaCert: &sqladmin.SslCert{Cert: certPEM.String()},
+	}
+
+	r := &Request{
+		reqMethod: http.MethodGet,
+		reqPath:   fmt.Sprintf("/sql/v1beta4/projects/%s/instances/%s", i.project, i.name),
+		reqCt:     ct,
+		handle: func(resp http.ResponseWriter, req *http.Request) {
+			b, err := db.MarshalJSON()
+			if err != nil {
+				http.Error(resp, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			resp.WriteHeader(http.StatusOK)
+			resp.Write(b)
+		},
+	}
+	return r
+}
+
+// CreateEphemeralSuccess returns a MockRequest that responds to the
+// `sslCerts.createEphemeral` SQL Admin endpoint. It responds with a "StatusOK" and a
+// SslCerts object.
+//
+// https://cloud.google.com/sql/docs/mysql/admin-api/rest/v1beta4/sslCerts/createEphemeral
+func CreateEphemeralSuccess(i CloudSQLInst, ct int) *Request {
+	r := &Request{
+		reqMethod: http.MethodPost,
+		reqPath:   fmt.Sprintf("/sql/v1beta4/projects/%s/instances/%s/createEphemeral", i.project, i.name),
+		reqCt:     ct,
+		handle: func(resp http.ResponseWriter, req *http.Request) {
+			// Read the body from the request.
+			b, err := ioutil.ReadAll(req.Body)
+			defer req.Body.Close()
+			if err != nil {
+				http.Error(resp, fmt.Errorf("unable to read body: %w", err).Error(), http.StatusBadRequest)
+				return
+			}
+			var eR sqladmin.SslCertsCreateEphemeralRequest
+			err = json.Unmarshal(b, &eR)
+			if err != nil {
+				http.Error(resp, fmt.Errorf("invalid or unexpected json: %w", err).Error(), http.StatusBadRequest)
+				return
+			}
+			// Extract the certificate from the request.
+			bl, _ := pem.Decode([]byte(eR.PublicKey))
+			if bl == nil {
+				http.Error(resp, fmt.Errorf("unable to decode PublicKey: %w", err).Error(), http.StatusBadRequest)
+				return
+			}
+			pubKey, err := x509.ParsePKIXPublicKey(bl.Bytes)
+			if err != nil {
+				http.Error(resp, fmt.Errorf("unable to decode PublicKey: %w", err).Error(), http.StatusBadRequest)
+				return
+			}
+
+			// Create a signed cert from the requests public key.
+			cert := &x509.Certificate{ // TODO: Validate this format vs API
```

Comment:
How much of these data is actually necessary for the tests? Could we omit any of it?